### DISCUSSION TITLE
Error: Plugin(NotFound("wasi_nn")), is it possible to bypass wasi-nn to perform ML?

### DISCUSSION THREAD

**Discussion by KeeganJin** [WasmEdge/WasmEdge#2809] [Q&A] [answered]
Hi all, 

I am following this [objection-detection_via_wasinn](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/object-detection-via-wasinn/src/main.rs) and this [wasmedge-bindgen](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/object-detection-via-wasinn/src/main.rs) example in [wasmedge-rustsdk-examples](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/object-detection-via-wasinn/src/main.rs) to build up a program which can call the **infer()** function in a compiled wasm file from the host function. 

However, when I was going through this [objection-detection_via_wasinn](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/object-detection-via-wasinn/src/main.rs)  examples, I met a Error: Plugin(NotFound("wasi_nn")).

Then, I manually set the path as where I download and export the plugins based on the instructions since the default setting is:
` // load wasinn-pytorch-plugin from the default plugin directory: /usr/local/lib/wasmedge`

However, still same error: Error: Plugin(NotFound("wasi_nn")).

Then I uninstalled WasmEdge and tried to re-install it for all users, following the instructions: [https://wasmedge.org/docs/start/install/#install-wasmedge-plug-ins-and-dependencies](url), but it shows 
`PermissionError: [Errno 13] Permission denied: 'libwasmedge.so.0.0.3' -> '/usr/local/lib/libwasmedge.so.0'` even though there is no such a file. 

so is there a way to fix that pluginManager issue?

Btw, is there a way to bypass wasi-nn to perform ML in WASM? It seems possible with the TF and TF-lite extensions.

I think it will be super nice to have a better document for WasmEdge, especially regarding how to set up the VM, can I help with that?

Thanks in advance!

**Comment by hydai** [comment]
I think the [objection-detection_via_wasinn](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/object-detection-via-wasinn/src/main.rs) is outdated. The plugins are handled by the installer now.
After users set up the NN frameworks by themselves, such as by installing PyTorch/OpenVINO, all they need to do is run the installer with `--plugins wasi_nn-pytorch`

Ref: https://wasmedge.org/docs/start/install/#wasi-nn-plug-in


/cc @apepkuss

**Reply by KeeganJin** [reply]
Hi, thanks for your quick reply.
I followed the instructions [https://wasmedge.org/docs/start/install/#wasi-nn-plug-in](url) and tested the plugin with this example [https://github.com/second-state/WasmEdge-WASINN-examples/tree/master/pytorch-mobilenet-image](url). This test without building up a VM works fine. But with [objection-detection_via_wasinn](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/object-detection-via-wasinn/src/main.rs), the error now changes to 
`thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', src/main.rs:20:24
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: core::panicking::panic_bounds_check
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:162:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/slice/index.rs:261:10
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/slice/index.rs:19:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/vec/mod.rs:2691:9
   6: object_detection_via_wasinn::infer
             at ./src/main.rs:20:24
   7: object_detection_via_wasinn::main
             at ./src/main.rs:11:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with RUST_BACKTRACE=full for a verbose backtrace.`

**Platform: ubuntu 22.**

I think, this is the only example that shows how to build up the VM for executing ML tasks in WASM, and it doesn't work currently :<. 
Can anyone help with that? Or maybe illustrate the right steps for building(setting up) the VM for executing ML tasks in a WASM? That would be super helpful. 
Thanks in advance.

**Comment by apepkuss** [comment, marked as answer]
Hi @KeeganJin, 

The [release](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/.github/workflows/release.yml) CI workflow could help you:

1. [Line 36 - 45](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/.github/workflows/release.yml#L36-L45) show how to install `libtorch`.
2. [Line 47 - 54](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/.github/workflows/release.yml#L47-L54) show how to install WasmEdge with wasi-nn pytorch plugin.
3. [Line 124 - 129](https://github.com/second-state/wasmedge-rustsdk-examples/blob/main/.github/workflows/release.yml#L124-L129) show how to run the `object-detection-via-wasinn` example.
